Add table for node addresses

joamaki opened the pull request:
To provide modules access to the evolving set of local node's addresses,                                                                                                                       
add Table[NodeAddress] that derives from the low-level Table[*Device] and                                                                                                                      
applies the Cilium-specific heuristics to pick which addresses are considered                                                                                                                  
host IPs and which are used for NodePort and BPF masquerading.                                                                                                                                 
                                                                                                                                                                                                
To allow user to expand the set of addresses used for NodePort, add the                                                                                                                        
configuration flag "--nodeport-addresses" for specifying from which CIDRs                                                                                                                      
the NodePort addresses are allowed. This mirrors exactly the same kube-proxy                                                                                                                   
flag. If user does not specify this, then the default behavior remains, which                                                                                                                  
is to pick the first IPv4 and IPv6 address of each native device.                                                                                                                              

```
// NodeAddress is an IP address assigned to a network interface on a Cilium node                                                                                                               
 // that is considered a "host" IP address.                                                                                                                                                     
 type NodeAddress struct {                                                                                                                                                                      
         Addr netip.Addr                                                                                                                                                                        
                                                                                                                                                                                                
         // NodePort is true if this address is to be used for NodePort.                                                                                                                        
         // If --nodeport-addresses is set, then all addresses on native                                                                                                                        
         // devices that are contained within the specified CIDRs are chosen.                                                                                                                   
         // If it is not set, then only the primary IPv4 and/or IPv6 address                                                                                                                    
         // of each native device is used.                                                                                                                                                      
         NodePort bool                                                                                                                                                                          
                                                                                                                                                                                                
         // Primary is true if this is the primary IPv4 or IPv6 address of this device.                                                                                                         
         // This is mainly used to pick the address for BPF masquerading.                                                                                                                       
         Primary bool                                                                                                                                                                           
                                                                                                                                                                                                
         // DeviceName is the name of the network device from which this address                                                                                                                
         // is derived from.                                                                                                                                                                    
         DeviceName string                                                                                                                                                                      
 } 
```

This PR is one of several that aims to replace the global maps in `pkg/node/address.go` containing the NodePort and BPF masquerade addresses with a table that can be watched for changes and allow dynamic reconfiguration when the addresses
change. The next PR will adapt `NodeAddressing` to access `Table[NodeAddress]` and removes the global maps from `pkg/node`.
Further PRs down the line will switch from `NodeAddressing` to `Table[NodeAddress]` and start reconciling on changes.

joamaki committed:
datapath/tables: Add Table[NodeAddress]

To provide modules access to the evolving set of local node's addresses,
add Table[NodeAddress] that derives from the low-level Table[*Device] and
applies the Cilium-specific heuristics to pick which addresses are considered
host IPs and which are used for NodePort and BPF masquerading.

To allow user to expand the set of addresses used for NodePort, add the
configuration flag "--nodeport-addresses" for specifying from which CIDRs
the NodePort addresses are allowed. This mirrors exactly the same kube-proxy
flag. If user does not specify this, then the default behavior remains, which
is to pick the first IPv4 and IPv6 address of each native device.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
hive: Add support for CIDR configuration flags

Add support for "*cidr.CIDR" and "[]*cidr.CIDR" in config
cells:

  type ExampleConfig {
    ExampleCIDRs []*cidr.CIDR
  }

  func (def ExampleConfig) Flags(flags *pflag.FlagSet) {
    flags.StringSlice("example-cidrs", def.ExampleCIDRs, "Set the example CIDRs")
  }
  // --example-cidrs=1.2.3.4/24,2001:db8::/64

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
datapath/devices: Add Secondary flag to DeviceAddress

Add 'Secondary' to 'DeviceAddress' struct to allow readers to sort
the addresses by scope & secondary when picking relevant addresses.
'Secondary' is the same as IF_A_SECONDARY.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/datapath/tables/node_address.go:
```diff
@@ -0,0 +1,365 @@
+	nodeAddressControllerMinInterval = 100 * time.Millisecond
+)
+
+type AddressScopeMax uint8
```
What's an `AddressScopeMax`?

joamaki replied:
Documented.

bimmlerd commented on pkg/datapath/tables/node_address.go:
```diff
@@ -0,0 +1,365 @@
+	return AddressScopeMax(daemonCfg.AddressScopeMax), nil
+}
+
+func (cfg NodeAddressConfig) getNets() []*net.IPNet {
```
Does this have to be old-style `net.IPNet` or can it be something from `netip`? Given this is new code, I'm hoping for the latter.

joamaki replied:
It's `net.IPNet` for `ip.NetsContainsAny`. Could add a `netip.Addr` variant for `NetsContainsAny` perhaps? Or do you have other ideas?

bimmlerd replied:
Yeah I think that would be the best option; it's unfortunate code duplication but we need to move towards that future..

bimmlerd commented on pkg/datapath/tables/node_address_test.go:
```diff
@@ -0,0 +1,453 @@
+					nodePort = append(nodePort, addr.Addr.String())
+				}
+			}
+			require.ElementsMatch(t, local, ipStrings(tt.wantLocal), "LocalAddresses do not match")
```
nit: these should be asserts, it's what the test is testing, right?

joamaki replied:
True, here it makes sense to assert and make sure to also stop!

bimmlerd commented on pkg/datapath/tables/node_address_test.go:
```diff
@@ -0,0 +1,453 @@
+	},
+}
+
+func TestNodeAddress(t *testing.T) {
```
nit: not that it matters much, but `t.Parallel()`?

bimmlerd commented on pkg/datapath/tables/node_address.go:
```diff
@@ -0,0 +1,365 @@
+		// We keep the scope-based address filtering as was introduced
+		// in 080857bdedca67d58ec39f8f96c5f38b22f6dc0b.
+		if addr.Scope > uint8(n.AddressScopeMax) {
+			// Addresses are sorted by scope, so can stop early.
```
but they're sorted by primary/secondary first, so a primary addr with larger scope would break but should not, afaict?

joamaki replied:
Ahh nice catch! Changing it to `continue` instead.

bimmlerd commented on pkg/datapath/tables/node_address.go:
```diff
@@ -0,0 +1,365 @@
+
+// sortAddresses returns a copy of the addresses, sorted by primary (e.g. !iIFA_F_SECONDARY) and then by
+// address scope.
+func sortAddresses(addrs []DeviceAddress) []DeviceAddress {
```
nit: mildly surprising to see a departure from the typical go sorting interface. It's called out in the comment though, so I'm okay with it, but maybe you can think of a way to work `copy` into the name of the function too?

joamaki replied:
How about if I change it to `sortedAddresses` (bit of a scala/java naming convention). Does that work?

joamaki replied:
Hmm, this also made me realize one of the big pitfalls of using slices in structs that are stored in statedb. Once there's func iters we might want to add `ImmutableList[T]` or something similar.

bimmlerd commented on pkg/datapath/tables/node_address.go:
```diff
@@ -0,0 +1,365 @@
+		skip := false
+		for _, prefix := range defaults.ExcludedDevicePrefixes {
+			if strings.HasPrefix(dev.Name, prefix) {
+				skip = true
```
why not just `return` here?

joamaki replied:
Hmm that's weird. Probably I had written this originally in a different style where there was `if skip { break }` etc. Fixed.

bimmlerd commented on pkg/datapath/tables/node_address.go:
```diff
@@ -0,0 +1,365 @@
+			// The user has not specified IP ranges to filter on IPs on which to serve NodePort.
+			// Thus the default behavior is to use the primary IPv4 and IPv6 addresses of each
+			// device.
+			if addr.Addr.Is4() && !ipv4 {
```
nit: had to double back and re-read the comment over `ipv4`, as it reads somewhat confusingly. Maybe `ipv4Found`/`ipv4Picked` instead?

joamaki replied:
I struggled finding good names/pattern here and not too happy how it looks right now, but didn't figure out a better way. `ipv4Found` is definitely an approvement so going with that.

bimmlerd reviewed (changes requested):
I think there's a bug when there's a primary address with a scope > scopeMax, followed by secondary addresses which have smaller scope. But I might just be missing context on what that scope means and that might be an impossible case.

gandro commented on pkg/datapath/tables/node_address.go:
```diff
@@ -0,0 +1,365 @@
+	limiter := rate.NewLimiter(nodeAddressControllerMinInterval, 1)
+	revision := statedb.Revision(0)
+	for {
+		txn := n.DB.WriteTxn(n.NodeAddresses)
```
Why does this transaction not need to lock n.Devices for the delete tracker?

joamaki replied:
WriteTxn on devices is required only when adding the delete tracker (it's added to the `tableEntry` in the radix tree of tables, hence we need to lock the table to modify it). It's not needed when reading since consuming deleted objects only requires an atomic set of the revision watermark on the specific delete tracker (this watermark is read by a background garbage collector which then does the actual deletions with a write transaction).

If you're interested in details check `(*txn).addDeleteTracker` and `pkg/statedb/graveyard.go` (garbage collector is here).

gandro replied:
Great, thanks for the pointers!

gandro commented on pkg/datapath/tables/node_address.go:
```diff
@@ -0,0 +1,365 @@
+		txn := n.DB.WriteTxn(n.NodeAddresses)
+		process := func(dev *Device, deleted bool, rev statedb.Revision) error {
+			addrIter, _ := n.NodeAddresses.Get(txn, NodeAddressDeviceNameIndex.Query(dev.Name))
+			existing := sets.New[NodeAddress](statedb.Collect[NodeAddress](addrIter)...)
```
non-blocking nit: A `statedb.CollectSet` (and perhaps also `statedb.CollectMap`) helper would be nice to avoid having to go through a intermediate slice

joamaki replied:
Good idea, I'll add it.

gandro commented on pkg/datapath/tables/node_address.go:
```diff
@@ -0,0 +1,365 @@
+	for addr := range new {
+		if !existing.Has(addr) {
+			updated = true
+			n.NodeAddresses.Insert(txn, addr)
```
can we `break` after the insertion here, given that `addr` should be the primary key?

joamaki replied:
No because we need to add all new addresses. Were just checking first if the address already exists to avoid mutation of the table and thus unnecessary triggering of watchers.

gandro reviewed (commented):
Looks good overall, I think having a table for this will be useful. I'm neither an expert on statedb, nor on node addressing

gandro reviewed (approved):
LGTM! (as mentioned before however, I'm a low-confidence reviewer both for statedb and node addresses)

gentoo-root commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -362,7 +362,8 @@ func (dc *devicesController) processUpdates(
 	return tables.DeviceAddress{
-		Addr: ip.MustAddrFromIP(upd.LinkAddress.IP),
+		Addr:      ip.MustAddrFromIP(upd.LinkAddress.IP),
+		Secondary: upd.Flags&unix.IFA_F_SECONDARY == 0,
```
Why `== 0`?

joamaki replied:
Ah had fixed this already in the original branch I forked this from, but looks like that change didn't carry over (this happened because this was originally `Primary` and didn't fix this line). Fixing this again and making sure to have the test coverage for it.

EDIT: Uh oh. I did have a test for this, but I had broken the test suite with the addition of the `orphanRoutes` check. Fixed that and `Secondary`.

joamaki committed:
datapath/devices: Fix selection on route changes and orphanRoutes check

The addition of the check to see if routes exists without corresponding
device to the devices controller tests caused it to stop the test too
early before the test case had passed.

As the tests were faulty they didn't catch the bug where the 'Selected'
status wasn't rechecked and thus the 'veth-with-default-route' test
passed when it shouldn't have. Fix this by always checking if selected
status updates.

Fixes: 60fd85aefd ("devices: Flush routes on device delete")
Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd reviewed (approved):
LGTM now! I think the netip stuff would be nice, but won't block the PR on it.
